### ISSUE TITLE
GPGPUProgram does not support complex64 input

### ISSUE THREAD

**Issue by JPery** [tensorflow/tfjs#4507]
**System information**

- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 20.04
- TensorFlow.js installed from (npm or script link): script link "//cdn.jsdelivr.net/npm/@tensorflow/tfjs@2.8.2/dist/tf.min.js"
- TensorFlow.js version (use command below): 2.8.2
- Tensorflow.js Converter Version: TensorFlow.js Converter v2.8.2
- Browser version: Google Chrome 83.0.4103.61

**Describe the current behavior**

- model.predict throws the following error:

```
Uncaught (in promise) Error: GPGPUProgram does not support complex64 input. For complex64 dtypes, please separate the program into real and imaginary parts.
    at backend_webgl.js:792
    at Array.map (<anonymous>)
    at t.n.runWebGLProgram (backend_webgl.js:790)
    at DG (Transpose_impl.js:30)
    at Object.OG [as kernelFunc] (Transpose.js:52)
    at p (engine.js:590)
    at engine.js:660
    at e.t.scopedRun (engine.js:453)
    at e.t.runKernelFunc (engine.js:657)
    at e.t.runKernel (engine.js:522)
```

**Describe the expected behavior**
It should run as normal and return the prediction

**Comment by pyu10055** [comment]
@JPery can you share the original saved model and converted model in order for us to reproduce. thanks.

**Comment by JPery** [comment]
Hi! 
I've uploaded the two models. [model-tf.zip](https://github.com/tensorflow/tfjs/files/5794742/model-tf.zip) and [model-tfjs.zip](https://github.com/tensorflow/tfjs/files/5794743/model-tfjs.zip).

I'm testing the tfjs one with  ```tf.zeros([1, 88200, 1])``` but it throws the error. When I load the model in Python with ```tf.saved_model.load``` and predict with ```result = saved_model(tf.zeros([20, 88200, 1]))``` no error is shown and it works as intended

**Comment by pyu10055** [comment]
@JPery Thank you for providing the models, it is the case that our WebGL backend does not support Complex tensors as input to any kernels, even though they can be outputs of kernels (i.e. rfft). One possible work around is in your python code, try to separate the real and img part of the complex tensors and perform follow up computation separately.

**Comment by JPery** [comment]
Hi, @pyu10055!

Couldn't this task be automated somehow? For example, converting this complex type to real and imaginary in runtime and then back or doing this transformation in the tfjs converter?

As @allanpichardo says, it will rule out some operations needed for some tasks.

Thanks for your answer!

**Comment by JPery** [comment]
Hi, @pyu10055. I've been trying to port my code to avoid the use of complex types.

In my case, x is the complex64 type. The code that throws the "GPGPUProgram does not support complex64 input" was:
```
tf.abs(x)
```

I'm changing it to:

```
tf.math.sqrt(
  tf.math.add(
      tf.math.multiply(tf.math.real(x), tf.math.real(x)),
      tf.math.multiply(tf.math.imag(x), tf.math.imag(x))
  )
)
```

but as `tf.math.real` and `tf.math.imag` are also tensors with complex64 input, I'm unable to run the code in tfjs. What can I do to transform my tensor to its real/imaginary parts without a tensor? Is there any other way to get the magnitude of the complex tensor?

Best regards!

**Comment by JPery** [comment]
Hi, @pyu10055!

Any updates on this?

Thanks!

**Comment by pyu10055** [comment]
@JPery tfjs tf.real and tf.imag ops do support complex tensor inputs, are you having trouble in converting the TF model to TFJS?

Can you share your errors, thanks.

**Comment by JPery** [comment]
Hi, @pyu10055! 

Thanks for your answer. I was getting the same "GPGPUProgram does not support complex64 input" but it was my bad, I wasn't converting the model properly.

Now, after changing the implementation of tf.abs for the other one mentioned above I still getting an error:

```
tensor_list.js:219 Uncaught (in promise) Error: Invalid data types; op elements complex64, but list elements null
    at TensorList.setItem (tensor_list.js:219)
    at _callee2$ (control_executor.js:241)
    at tryCatch (runtime.js:45)
    at Generator.invoke [as _invoke] (runtime.js:274)
    at Generator.prototype.<computed> [as next] (runtime.js:97)
    at asyncGeneratorStep (runtime.js:728)
    at _next (runtime.js:728)
    at runtime.js:728
    at new Promise (<anonymous>)
    at runtime.js:728
setItem @ tensor_list.js:218
_callee2$ @ control_executor.js:28
tryCatch @ runtime.js:45
invoke @ runtime.js:272
prototype.<computed> @ runtime.js:96
asyncGeneratorStep @ runtime.js:728
_next @ runtime.js:728
(anonymous) @ runtime.js:728
(anonymous) @ runtime.js:728
executeOp @ control_executor.js:28
(anonymous) @ operation_executor.js:64
executeOp$h @ operation_executor.js:57
_loop @ graph_executor.js:491
processStack @ graph_executor.js:477
_callee4$ @ graph_executor.js:404
tryCatch @ runtime.js:45
invoke @ runtime.js:272
prototype.<computed> @ runtime.js:96
asyncGeneratorStep @ runtime.js:728
_next @ runtime.js:728
```

The fact is that I'm trying to convert a Keras model with custom layers to a tf_saved_model using ```tf.saved_model.save``` (this part works well even when I load the pb model in Python) in order to avoid the implementation of the custom layers in TFJS. I don't know if it could cause any additional issue.

**Comment by pyu10055** [comment]
@JPery Looks like there are still complex64 type in the model, especially used on TensorList ops. i.e. ("StatefulPartitionedCall/sequential/melspectrogram/stft/stft_tf.signal.stft/map/TensorArrayV2Stack/TensorListStack")
It seems to related to melspectrogram/stft layer or method?

**Comment by JPery** [comment]
Hi!

I finally found my error. In the custom layer I was using there was a tf.transpose right after the tf.signal.stft. This operation was the one causing the issue, as tf.abs has a complex64 input implementation in tfjs.

In the end, I moved the tf.abs right before the tf.transpose (it was right after) and it resolved the issue.

On the other hand, could we get a tf.transpose complex64 input implementation? 

Thanks a lot for your help, @pyu10055
